mutex.go

```diff
@@ -44,10 +44,21 @@ var (
 // Lock waits up to duration d for l.Lock() to succeed.
 func Lock(l sync.Locker, d time.Duration) error {
 	done := make(chan struct{}, 1)
-	go func() {
-		l.Lock()
-		done <- struct{}{}
-	}()
+	m, ok := l.(*mutex)
+	// if l is a mutex, lock with a timeout context.
+	if ok {
+		ctx, cancel := context.WithTimeout(context.Background(), d*time.Second)
```

Comment:
I wrote "next commit" before I submitted it in case you checked and saw my comment and wondered what I was talking about :). I'm sure about the correct protocol between claiming a change is done and submitting the associated commit (w/ Gerrit, you explicitly indicate when a set of comments should be shared).